dvc/repo/checkout.py

```diff
@@ -58,6 +58,7 @@ def checkout(
         "deleted": [],
         "modified": [],
         "failed": [],
+        "failed_not_stored": [],
```

Comment:
I don't think we should introduce new status. What matters from the point of view of `checkout` is that particular file failed. We should not spread logic related to `import` into `checkout`. Also, if we were to go with that status, we would need to revisit all situations, where remote does not contain particular hash (eg forgot to push) and in that case `failed_not_stored` would probably be more appropriate than just `failed`.

Reply:
I wrote a bit below, but I'd point out that the logic here isn't logically linked to `import`: `store: false` is a property of an `out`, which just happens to be set by `import`. 

Also, `failed_not_stored` above wasn't mean to mean "these files failed because they aren't stored on the remote", it's meant to mean "these files failed and they have `store: false` set". I agree it's a little unsatisfying, but I don't think checkout is really being coupled to the import logic here: it's being linked to the settings of an `out`, but that's reasonable since that's what `checkout` is concerned with.

Reply:
You are right, I exaggerated here, the thing is about stage and outs rather than `import`.
Having said so, `failed_not_stored` is a piece of information connecting `stage` with `checkout`. 
It is the context for the reason why checkout for particular stage failed. The thing about `checkout` is that it is supposed to give summarized information about the state of the cache. So, till today we have been having 4 basic "stats" - which aggregate all possible outcomes of checkout.

Introducing hinting as an integral part of checkout is precedence that might result in more and more stats fields: in most cases, we can pinpoint why the checkout failed - and suggest user what is the problem. In long term, this approach is unmaintainable, since there will be a lot of use cases and reasons to fail.  

For example: what if in the current state of change, the output will fail to checkout for a different reason, for example, access constraints? Wouldn't it mean that the error would still point to out being not stored if `out.store` is true?

Reply:
> For example: what if in the current state of change, the output will fail to checkout for a different reason, for example, access constraints? Wouldn't it mean that the error would still point to out being not stored if out.store is true?

Hmm yes, you're right about that. It's a hint rather than an error, but still. 

I also completely get your point about maintainability if we try to hint for every possible failure mode. 

That said, I do think this failure mode is a bit special, since it results from the use of a feature entirely within dvc (i.e. "store: false"). Compared to e.g. an access constraint or a network failure which are external failures, this one is due to settings which we have complete knowledge about. 

I'm not strongly in favor of keeping the hint, maybe a logger warning (or info?) is the easiest way around the problem.

Reply:
> It's a hint rather than an error, but still.

This is a problem about this approach: we would introduce new `checkout` status, in order to be able to hint user, knowing its just a hunch, while the actual reason for failure might be different. I totally get that this error comes "from design" and its detection is a bit different than other errors. I just think remodeling whole `checkout` in order to give user a hint what is wrong is a bit too much. 

Don't get me wrong, I am all for hinting, and we actually have been doing it quite a lot at a time, but introducing multiple hints over time results in a lot of `WARNING` log entries even in case of simple use cases, which gets users confused. 

That being said, in case of more advanced users, it would be desirable to be able to point them to potential solution. And that is why I suggested `logger.debug` on checkout - its a hint, does not require major changes in code and can be implemented in single place, the actual origin of `CheckoutError`. I guess it makes sense to make it into `warning`, as this after checkout, user won't be able to reproduce the results (since there will be no data to check out).

Reply:
Alright, I'm convinced :) I'll go ahead and revert the hinting stuff.